File: lib/keypoints/keypoint_normalizer.dart
import 'dart:math' as math;
import 'package:ishara/keypoints/ishara_keypoint_mapper.dart';

/// KeypointNormalizer
/// التطبيع الرياضي الدقيق المطابق حرفياً لكود التدريب datasetv2.py / PoseDatasetV2
/// يطبّق نفس المعادلات على المجموعات الـ 4:
/// 1. اليد اليمنى (21 نقطة)
/// 2. اليد اليسرى (21 نقطة)
/// 3. الشفاه (19 نقطة)
/// 4. الجسم والرأس (25 نقطة)
class KeypointNormalizer {
  /// تطبيع مجموعة نقاط ثنائية الأبعاد [N, 2] حسب كود التدريب:
  /// 1. طرح أول نقطة (pose[0]) لجعل المرجع المعصم/الشفة/العنق
  /// 2. طرح أقل قيمة في كل محور لجعل البداية من الصفر
  /// 3. القسمة على أكبر قيمة في المحورين للتحجيم داخل صندوق 1x1
  /// 4. طرح المتوسط العام لكل العناصر
  /// 5. القسمة على القيمة المطلقة العظمى والضرب في 0.5 للتوسيط داخل [-0.5, 0.5]
  static List<List<double>> normalizeSubset(List<List<double>> subset) {
    final int n = subset.length;
    if (n == 0) return [];

    // التحقق هل جميع النقاط أصفار (غير مكتشفة)
    double sum = 0.0;
    for (int i = 0; i < n; i++) {
      sum += subset[i][0].abs() + subset[i][1].abs();
    }
    if (sum < 1e-6) {
      return List.generate(n, (_) => [0.0, 0.0]);
    }

    // إنشاء نسخة للتعديل
    final result = List.generate(n, (i) => [subset[i][0], subset[i][1]]);

    // 1. طرح أول نقطة pose[0]
    final double originX = result[0][0];
    final double originY = result[0][1];
    for (int i = 0; i < n; i++) {
      result[i][0] -= originX;
      result[i][1] -= originY;
    }

    // 2. طرح أقل قيمة في كل محور
    double minX = result[0][0];
    double minY = result[0][1];
    for (int i = 1; i < n; i++) {
      if (result[i][0] < minX) minX = result[i][0];
      if (result[i][1] < minY) minY = result[i][1];
    }
    for (int i = 0; i < n; i++) {
      result[i][0] -= minX;
      result[i][1] -= minY;
    }

    // 3. التحجيم في صندوق 1x1 (max_vals = np.max(pose, axis=0); pose /= max(max_vals))
    double maxX = result[0][0];
    double maxY = result[0][1];
    for (int i = 1; i < n; i++) {
      if (result[i][0] > maxX) maxX = result[i][0];
      if (result[i][1] > maxY) maxY = result[i][1];
    }
    final double maxCoord = math.max(maxX, maxY);
    if (maxCoord > 1e-7) {
      for (int i = 0; i < n; i++) {
        result[i][0] /= maxCoord;
        result[i][1] /= maxCoord;
      }
    }

    // 4. طرح المتوسط العام لجميع العناصر (np.mean(pose[:, :]))
    double totalVal = 0.0;
    for (int i = 0; i < n; i++) {
      totalVal += result[i][0] + result[i][1];
    }
    final double meanVal = totalVal / (2.0 * n);
    for (int i = 0; i < n; i++) {
      result[i][0] -= meanVal;
      result[i][1] -= meanVal;
    }

    // 5. القسمة على القيمة المطلقة العظمى والضرب في 0.5 (np.max(np.abs(pose[:, :])) * 0.5)
    double maxAbs = 0.0;
    for (int i = 0; i < n; i++) {
      final absX = result[i][0].abs();
      final absY = result[i][1].abs();
      if (absX > maxAbs) maxAbs = absX;
      if (absY > maxAbs) maxAbs = absY;
    }

    if (maxAbs > 1e-7) {
      for (int i = 0; i < n; i++) {
        result[i][0] = (result[i][0] / maxAbs) * 0.5;
        result[i][1] = (result[i][1] / maxAbs) * 0.5;
      }
    }

    return result;
  }

  /// تطبيع إطار كامل من 86 نقطة مع الحفاظ على النقاط المفقودة أو استراتيجية Carry-forward
  static KeypointFrame normalizeForModel(
    KeypointFrame rawFrame, {
    KeypointFrame? previousFrame,
  }) {
    final rawKeypoints = rawFrame.keypoints;
    final List<ModelKeypointInfo> normalizedList = [];

    // تقسيم المجموعات الـ 4
    List<List<double>> extractSubset(int start, int count) {
      final List<List<double>> sub = [];
      for (int i = 0; i < count; i++) {
        final kp = rawKeypoints[start + i];
        if (kp.isValid) {
          sub.add([kp.x!, kp.y!]);
        } else {
          // استراتيجية التدريب في datasetv2.py:
          // في حال فقدان نقاط اليد أو الجسم، يتم أخذ آخر إطار سليم (Carry-forward)
          // وإن لم يوجد، تُعطى 0.0
          if (previousFrame != null &&
              previousFrame.keypoints[start + i].isValid) {
            sub.add([
              previousFrame.keypoints[start + i].x!,
              previousFrame.keypoints[start + i].y!,
            ]);
          } else {
            sub.add([0.0, 0.0]);
          }
        }
      }
      return sub;
    }

    // 1. Right Hand [0..20]
    final rhRaw = extractSubset(0, IsharaKeypointMapper.numRightHand);
    final rhNorm = normalizeSubset(rhRaw);
    for (int i = 0; i < IsharaKeypointMapper.numRightHand; i++) {
      final isRawValid = rawKeypoints[i].isValid;
      normalizedList.add(rawKeypoints[i].copyWithCoordinates(
        isRawValid ? rhNorm[i][0] : null,
        isRawValid ? rhNorm[i][1] : null,
      ));
    }

    // 2. Left Hand [21..41]
    final lhRaw = extractSubset(21, IsharaKeypointMapper.numLeftHand);
    final lhNorm = normalizeSubset(lhRaw);
    for (int i = 0; i < IsharaKeypointMapper.numLeftHand; i++) {
      final isRawValid = rawKeypoints[21 + i].isValid;
      normalizedList.add(rawKeypoints[21 + i].copyWithCoordinates(
        isRawValid ? lhNorm[i][0] : null,
        isRawValid ? lhNorm[i][1] : null,
      ));
    }

    // 3. Face/Lips [42..60]
    final fcRaw = extractSubset(42, IsharaKeypointMapper.numLips);
    final fcNorm = normalizeSubset(fcRaw);
    for (int i = 0; i < IsharaKeypointMapper.numLips; i++) {
      final isRawValid = rawKeypoints[42 + i].isValid;
      normalizedList.add(rawKeypoints[42 + i].copyWithCoordinates(
        isRawValid ? fcNorm[i][0] : null,
        isRawValid ? fcNorm[i][1] : null,
      ));
    }

    // 4. Body/Head [61..85]
    final bdRaw = extractSubset(61, IsharaKeypointMapper.numBody);
    final bdNorm = normalizeSubset(bdRaw);
    for (int i = 0; i < IsharaKeypointMapper.numBody; i++) {
      final isRawValid = rawKeypoints[61 + i].isValid;
      normalizedList.add(rawKeypoints[61 + i].copyWithCoordinates(
        isRawValid ? bdNorm[i][0] : null,
        isRawValid ? bdNorm[i][1] : null,
      ));
    }

    return KeypointFrame(
      keypoints: normalizedList,
      timestamp: rawFrame.timestamp,
      isTrainingMappingVerified: rawFrame.isTrainingMappingVerified,
    );
  }
}

File: android/app/src/main/kotlin/com/Ishara/iq/ishara/vision/KeypointNormalizer.kt
package com.Ishara.iq.ishara.vision

import kotlin.math.abs
import kotlin.math.max

/**
 * معالج ومطبع النقاط الـ 86 الرياضي (KeypointNormalizer)
 * مطابق حرفياً لكود التدريب الأصلي datasetv2.py / PoseDatasetV2
 * يطبّق نفس الخطوات الخمس على المجموعات الأربع:
 * 1. Right Hand (21)
 * 2. Left Hand (21)
 * 3. Lips (19)
 * 4. Body (25)
 */
object KeypointNormalizer {

    /**
     * تطبيع مجموعة فرعية من النقاط [N, 2] حسب datasetv2.py
     */
    fun normalizeSubset(subset: Array<FloatArray>): Array<FloatArray> {
        val n = subset.size
        if (n == 0) return emptyArray()

        // 1. التحقق هل جميع النقاط أصفار
        var sum = 0.0f
        for (i in 0 until n) {
            sum += abs(subset[i][0]) + abs(subset[i][1])
        }
        if (sum < 1e-6f) {
            return Array(n) { floatArrayOf(0.0f, 0.0f) }
        }

        // إنشاء نسخة للعمليات الحسابية
        val result = Array(n) { i -> floatArrayOf(subset[i][0], subset[i][1]) }

        // 1. طرح أول نقطة pose[0]
        val originX = result[0][0]
        val originY = result[0][1]
        for (i in 0 until n) {
            result[i][0] -= originX
            result[i][1] -= originY
        }

        // 2. طرح أقل قيمة في كل محور لجعل البداية من الصفر
        var minX = result[0][0]
        var minY = result[0][1]
        for (i in 1 until n) {
            if (result[i][0] < minX) minX = result[i][0]
            if (result[i][1] < minY) minY = result[i][1]
        }
        for (i in 0 until n) {
            result[i][0] -= minX
            result[i][1] -= minY
        }

        // 3. التحجيم في صندوق 1x1: max_coord = max(maxX, maxY); pose /= max_coord
        var maxX = result[0][0]
        var maxY = result[0][1]
        for (i in 1 until n) {
            if (result[i][0] > maxX) maxX = result[i][0]
            if (result[i][1] > maxY) maxY = result[i][1]
        }
        val maxCoord = max(maxX, maxY)
        if (maxCoord > 1e-7f) {
            for (i in 0 until n) {
                result[i][0] /= maxCoord
                result[i][1] /= maxCoord
            }
        }

        // 4. طرح المتوسط العام لجميع العناصر: mean = total / (2 * n)
        var totalVal = 0.0f
        for (i in 0 until n) {
            totalVal += result[i][0] + result[i][1]
        }
        val meanVal = totalVal / (2.0f * n)
        for (i in 0 until n) {
            result[i][0] -= meanVal
            result[i][1] -= meanVal
        }

        // 5. القسمة على القيمة المطلقة العظمى والضرب في 0.5: pose = (pose / maxAbs) * 0.5
        var maxAbs = 0.0f
        for (i in 0 until n) {
            val ax = abs(result[i][0])
            val ay = abs(result[i][1])
            if (ax > maxAbs) maxAbs = ax
            if (ay > maxAbs) maxAbs = ay
        }
        if (maxAbs > 1e-7f) {
            for (i in 0 until n) {
                result[i][0] = (result[i][0] / maxAbs) * 0.5f
                result[i][1] = (result[i][1] / maxAbs) * 0.5f
            }
        }

        return result
    }

    /**
     * تطبيع إطار كامل من 86 نقطة مع تطبيق استراتيجية Carry-forward للنقاط المفقودة
     */
    fun normalizeFrame(
        rawFrame: KeypointFrame86,
        previousNormalizedFrame: KeypointFrame86? = null
    ): KeypointFrame86 {
        val rawPoints = rawFrame.keypoints
        val normalizedPoints = Array(KeypointMapper86.TOTAL_KEYPOINTS) { Keypoint2D.MISSING }
        val imputedMask = BooleanArray(KeypointMapper86.TOTAL_KEYPOINTS) { false }

        fun extractAndNormalizeSubset(
            offset: Int,
            count: Int
        ) {
            val sub = Array(count) { i ->
                val kp = rawPoints[offset + i]
                if (kp.isValid) {
                    floatArrayOf(kp.x, kp.y)
                } else {
                    // Carry-forward من الإطار السابق إذا وجد، وإلا أصفار
                    val prevKp = previousNormalizedFrame?.keypoints?.get(offset + i)
                    if (prevKp != null && prevKp.isValid) {
                        imputedMask[offset + i] = true
                        floatArrayOf(prevKp.x, prevKp.y)
                    } else {
                        floatArrayOf(0.0f, 0.0f)
                    }
                }
            }

            val normed = normalizeSubset(sub)
            for (i in 0 until count) {
                val isRawValid = rawPoints[offset + i].isValid
                val isImputed = imputedMask[offset + i]
                val isValid = isRawValid || isImputed
                normalizedPoints[offset + i] = Keypoint2D(
                    x = normed[i][0],
                    y = normed[i][1],
                    isValid = isValid
                )
            }
        }

        // 1. Right Hand [0..20]
        extractAndNormalizeSubset(KeypointMapper86.RIGHT_HAND_OFFSET, KeypointMapper86.NUM_RIGHT_HAND)

        // 2. Left Hand [21..41]
        extractAndNormalizeSubset(KeypointMapper86.LEFT_HAND_OFFSET, KeypointMapper86.NUM_LEFT_HAND)

        // 3. Lips [42..60]
        extractAndNormalizeSubset(KeypointMapper86.LIPS_OFFSET, KeypointMapper86.NUM_LIPS)

        // 4. Body [61..85]
        extractAndNormalizeSubset(KeypointMapper86.BODY_OFFSET, KeypointMapper86.NUM_BODY)

        return KeypointFrame86(
            keypoints = normalizedPoints,
            timestampMs = rawFrame.timestampMs,
            rightHandValid = rawFrame.rightHandValid,
            leftHandValid = rawFrame.leftHandValid,
            lipsValid = rawFrame.lipsValid,
            bodyValid = rawFrame.bodyValid,
            imputedMask = imputedMask
        )
    }
}
